workspaces/libnpmpublish/lib/provenance.js

```diff
@@ -1,68 +1,204 @@
+      subject,
+      predicateType: SLSA_PREDICATE_TYPE,
+      predicate: {
+        buildType: `${GITHUB_BUILD_TYPE_PREFIX}/${GITHUB_BUILD_TYPE_VERSION}`,
+        builder: { id: GITHUB_BUILDER_ID },
+        invocation: {
+          configSource: {
+            uri: `git+${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}@${env.GITHUB_REF}`,
+            digest: {
+              sha1: env.GITHUB_SHA,
+            },
+            entryPoint: workflowPath,
+          },
+          parameters: {},
+          environment: {
+            GITHUB_EVENT_NAME: env.GITHUB_EVENT_NAME,
+            GITHUB_REF: env.GITHUB_REF,
+            GITHUB_REPOSITORY: env.GITHUB_REPOSITORY,
+            GITHUB_REPOSITORY_ID: env.GITHUB_REPOSITORY_ID,
+            GITHUB_REPOSITORY_OWNER_ID: env.GITHUB_REPOSITORY_OWNER_ID,
+            GITHUB_RUN_ATTEMPT: env.GITHUB_RUN_ATTEMPT,
+            GITHUB_RUN_ID: env.GITHUB_RUN_ID,
+            GITHUB_SHA: env.GITHUB_SHA,
+            GITHUB_WORKFLOW_REF: env.GITHUB_WORKFLOW_REF,
+            GITHUB_WORKFLOW_SHA: env.GITHUB_WORKFLOW_SHA,
           },
-          entryPoint: workflowPath,
         },
-        parameters: {},
-        environment: {
-          GITHUB_EVENT_NAME: env.GITHUB_EVENT_NAME,
-          GITHUB_REF: env.GITHUB_REF,
-          GITHUB_REPOSITORY: env.GITHUB_REPOSITORY,
-          GITHUB_REPOSITORY_ID: env.GITHUB_REPOSITORY_ID,
-          GITHUB_REPOSITORY_OWNER_ID: env.GITHUB_REPOSITORY_OWNER_ID,
-          GITHUB_RUN_ATTEMPT: env.GITHUB_RUN_ATTEMPT,
-          GITHUB_RUN_ID: env.GITHUB_RUN_ID,
-          GITHUB_SHA: env.GITHUB_SHA,
-          GITHUB_WORKFLOW_REF: env.GITHUB_WORKFLOW_REF,
-          GITHUB_WORKFLOW_SHA: env.GITHUB_WORKFLOW_SHA,
+        metadata: {
+          buildInvocationId: `${env.GITHUB_RUN_ID}-${env.GITHUB_RUN_ATTEMPT}`,
+          completeness: {
+            parameters: false,
+            environment: false,
+            materials: false,
+          },
+          reproducible: false,
         },
+        materials: [
+          {
+            uri: `git+${env.GITHUB_SERVER_URL}/${env.GITHUB_REPOSITORY}@${env.GITHUB_REF}`,
+            digest: {
+              sha1: env.GITHUB_SHA,
+            },
+          },
+        ],
       },
-      metadata: {
-        buildInvocationId: `${env.GITHUB_RUN_ID}-${env.GITHUB_RUN_ATTEMPT}`,
-        completeness: {
-          parameters: false,
-          environment: false,
-          materials: false,
+    }
+  }
+  if (ci.GITLAB) {
+    // TODO: pull values from authenticated JWT rather than environment variables.
+    payload = {
+      _type: INTOTO_STATEMENT_TYPE,
+      subject,
+      predicateType: SLSA_PREDICATE_TYPE,
+      predicate: {
+        buildType: `${GITLAB_BUILD_TYPE_PREFIX}/${GITLAB_BUILD_TYPE_VERSION}`,
+        builder: { id: `${env.CI_PROJECT_URL}/-/runners/${env.CI_RUNNER_ID}` },
+        invocation: {
+          configSource: {
+            uri: `git+${env.CI_PROJECT_URL}@${env.CI_BUILD_REF}`,
```

Comment:
@wlynch I'm trying to map these predicate values in `configSource` to the [Fulcio OID mappings](https://github.com/sigstore/fulcio/pull/1097/files) so we could look at validating the certificate extensions against what's in the predicate to make sure they haven't been forged/spoofed during build.

- Does `CI_PROJECT_URL` map to the `server_url + project_path` claims set as the `Source Repository URI`?
- Does `CI_BUILD_REF` map to the `ref` claim set as the `Source Repository Ref`?
- Does `CI_COMMIT_SHA` map to the `sha` claim set as the `Source Repository Digest`
- Does `CI_JOB_NAME ` map to `pipeline_ref` claim that's set as the `Build Config URI`? 
- Does `CI_JOB_URL` map to the `server_url + project_path + /-/pipelines/ + pipeline_id` claims set as the `Run Invocation URI`?

I realise changing some of these to map would mean the provenance statement is no longer in line with the GitLab v0.2 schema, but any thoughts on getting this changed to align with the Fulcio changes? We could punt on this consistency verification for GitLab and just pull these values from the cert for UI display but means someone could still easily mess with the build env, spoof details and someone inspecting the predicate on Rekor would see the incorrect values unless the also checked the cert and compared.

Reply:
> Does CI_PROJECT_URL map to the server_url + project_path claims set as the Source Repository URI?

✅  Yes.

> Does CI_BUILD_REF map to the ref claim set as the Source Repository Ref?

🟡  Mostly. In the cert this will be the ref path (e.g. the full `refs/heads/main` path).

> Does CI_COMMIT_SHA map to the sha claim set as the Source Repository Digest

✅ Yes

> Does CI_JOB_NAME  map to pipeline_ref claim that's set as the Build Config URI?

❌ No. This would be the name of the job in the pipeline, but jobs are only addressable by id, not name.

> Does CI_JOB_URL map to the server_url + project_path + /-/pipelines/ + pipeline_id claims set as the Run Invocation URI?

❌ No this would match the Build Config URI.

Reply:
Added additional fields in environment to capture pipeline ref + pipeline id.